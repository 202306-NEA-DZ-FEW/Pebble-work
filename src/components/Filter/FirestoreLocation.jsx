import React, { useState, useEffect, useRef } from "react";
import { db } from "@/util/firebase";
import { doc, getDoc, collection } from "firebase/firestore";
import styles from "@/styles/Events.module.css";
import { AiOutlineClose } from "react-icons/ai";

function FirestoreLocation({ onInputChange }) {
    const [filteredLocations, setFilteredLocations] = useState([]);
    const [searchTerm, setSearchTerm] = useState("");
    const [locationsArray, setLocationsArray] = useState([]);

    const handleLocationClick = (selectedLocation) => {
        onInputChange(selectedLocation);
        console.log(`Selected location: ${selectedLocation}`);
    };
    const handleSearchChange = (e) => {
        setSearchTerm(e.target.value.toLowerCase());
    };
    useEffect(() => {
        const filtered = locationsArray.filter((location) =>
            location.toLowerCase().includes(searchTerm)
        );
        setFilteredLocations(filtered);
    }, [searchTerm, locationsArray]);
    const fetchLocationData = async () => {
        try {
            const locationsDocRef = doc(db, "database", "locations");
            const locationsDoc = await getDoc(locationsDocRef);
            if (locationsDoc.exists()) {
                const locationData = locationsDoc.data();
                const locationsArray = [].concat(
                    ...Object.values(locationData)
                );
                setFilteredLocations(locationsArray);
                setLocationsArray(locationsArray);
            }
        } catch (error) {
            console.error("Error fetching locations:", error);
        }
    };

    useEffect(() => {
        fetchLocationData();
    }, []);

    return (
        <>
            <button
                className='btn sm:bg-blue-500   hover:bg-gray-400 sm:text-center flex lg:items-center sm:gap-0 gap-2 sm:justify-center lg:w-[281px] lg:h-[52px] sm:w-[149px] sm:h-[30px] text-white font-[500] lg:text-[18px] lg:tracking-[0.10px] sm:text-[12px] sm:px-4 sm:py-2 rounded-lg'
                onClick={() =>
                    document.getElementById("my_modal_1").showModal()
                }
            >
                Location
            </button>
            <dialog id='my_modal_1' className='modal rounded-tr rounded-br'>
                <div className='modal-box'>
                    <div className='flex justify-center'>
                        <input
                            type='text'
                            placeholder='Search city...'
                            value={searchTerm}
                            onChange={handleSearchChange}
                            className='mt-2 p-2 border rounded'
                        />
                    </div>
                    <h3 className='font-bold text-lg mb-0'>Citys </h3>
                    <div className=' mt-2  '>
                        <ul
                            className={`list-none ${styles.information} h-[200px]`}
                        >
                            {Object.entries(filteredLocations).map(
                                ([key, value], index) => (
                                    <li
                                        key={index}
                                        className='my-2 block hover:bg-gray-400 cursor-pointer'
                                        onClick={() => {
                                            handleLocationClick(value);

                                            document
                                                .getElementById("my_modal_1")
                                                .close();
                                        }}
                                    >
                                        {value}
                                    </li>
                                )
                            )}
                        </ul>

                        <div className='modal-action absolute top-7 right-3 mt-0'>
                            <form method='dialog'>
                                <button
                                    className=''
                                    onClick={() =>
                                        document
                                            .getElementById("my_modal_1")
                                            .close()
                                    }
                                >
                                    <AiOutlineClose size={20} />
                                </button>
                            </form>
                        </div>
                    </div>
                </div>
            </dialog>
        </>
    );
}

export default FirestoreLocation;
