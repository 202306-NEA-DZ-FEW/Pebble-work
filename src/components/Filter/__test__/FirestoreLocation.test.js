import FirestoreLocation from "../FirestoreLocation";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";

jest.mock("@/util/firebase", () => ({
    db: jest.fn(),
}));
jest.mock("firebase/firestore", () => ({
    doc: jest.fn(),
    getDoc: jest.fn(),
}));

describe("FirestoreLocation Component", () => {
    const mockProps = {
        onInputChange: jest.fn(),
        onInputDelete: jest.fn(),
        inputValue1: "",
        setInputValue1: jest.fn(),
    };

    const renderComponent = (props = {}) => {
        render(<FirestoreLocation {...mockProps} {...props} />);
    };

    test("should render component", () => {
        renderComponent();
        expect(renderComponent()).toMatchSnapshot();
    });
});
