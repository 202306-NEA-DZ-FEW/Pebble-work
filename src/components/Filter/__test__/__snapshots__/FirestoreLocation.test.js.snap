// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirestoreLocation Component should render component 1`] = `undefined`;
